api/hypershift/v1beta1/azure_types.go

```diff
@@ -0,0 +1,213 @@
+	// diskEncryptionSetID is the ID of the DiskEncryptionSet resource to use to encrypt the OS disks for the VMs. This
+	// needs to exist in the same subscription id listed in the Hosted Cluster, HostedCluster.Spec.Platform.Azure.SubscriptionID.
+	// DiskEncryptionSetID should also exist in a resource group under the same subscription id and the same location
+	// listed in the Hosted Cluster, HostedCluster.Spec.Platform.Azure.Location.
+	// TODO: What does this do for a customer, why would they want to set it?
+	// TODO: What is the valid character set for this field? What about minimum and maximum lengths?
+	//
+	// +optional
+	DiskEncryptionSetID string `json:"diskEncryptionSetID,omitempty"`
+
+	// TODO: This shouldn't be a bool, but I don't know what the field is really for.
+	// Should it be coming under the disk parameters? We need to know what the impact of this is for users.
+	// Is this a particularly bad idea for control plane nodes, and does that matter for HCP, probably not.
+	//
+	// enableEphemeralOSDisk is a flag when set to true, will enable ephemeral OS disk.
+	//
+	// +optional
+	EnableEphemeralOSDisk bool `json:"enableEphemeralOSDisk,omitempty"`
+
+	// subnetID is the subnet ID of an existing subnet where the nodes in the nodepool will be created. This can be a
+	// different subnet than the one listed in the HostedCluster, HostedCluster.Spec.Platform.Azure.SubnetID, but must
+	// exist in the same network, HostedCluster.Spec.Platform.Azure.VnetID, and must exist under the same subscription ID,
+	// HostedCluster.Spec.Platform.Azure.SubscriptionID.
+	// subnetID is immutable once set.
+	// TODO: What is the valid character set for this field? What about minimum and maximum lengths?
+	//
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="SubnetID is immutable"
+	// +kubebuilder:validation:Required
+	SubnetID string `json:"subnetID"`
+
+	// diagnostics specifies the diagnostics settings for a virtual machine.
+	// If not specified, then Boot diagnostics will be disabled.
+	// +optional
+	Diagnostics *Diagnostics `json:"diagnostics,omitempty"`
+
+	// machineIdentityID is a user-assigned identity assigned to the VMs used to authenticate with Azure services. The
+	// identify is expected to exist under the same resource group as HostedCluster.Spec.Platform.Azure.ResourceGroupName. This
+	// user assigned identity is expected to have the Contributor role assigned to it and scoped to the resource group
+	// under HostedCluster.Spec.Platform.Azure.ResourceGroupName.
+	//
+	// If this field is not supplied, the Service Principal credentials will be written to a file on the disk of each VM
+	// in order to be accessible by the cloud provider; the aforementioned credentials provided are the same ones as
+	// HostedCluster.Spec.Platform.Azure.Credentials. However, this is less secure than using a managed identity.
+	//
+	// TODO: What is the valid character set for this field? What about minimum and maximum lengths?
```

Comment:
So it follows this structure - https://github.com/openshift/hypershift/blob/b0a6b0af99a9120e77994907f0c702b5dd44980b/vendor/sigs.k8s.io/cluster-api-provider-azure/api/v1beta1/types.go#L568

From there its just checking each component in https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules I believe.

Reply:
Great, we can come up with a pattern to match for this

Reply:
Others don't always start with the Azure prefix, this one definitely does? Should we prefix that ourselves to avoid the difference?

Reply:
>If this field is not supplied, the Service Principal credentials will be written to a file on the disk of each VM

why do we want to support this?

Reply:
We don't have to, but I think from the conversations we've had, this is what happens today. You could make this field required and ensure that the machineIdentity is provided